app/src/main/java/free/rm/skytube/gui/businessobjects/CleanerDialog.java

```diff
@@ -0,0 +1,131 @@
+import free.rm.skytube.businessobjects.db.PlaybackStatusDb;
+import free.rm.skytube.gui.activities.MainActivity;
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
+import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.disposables.CompositeDisposable;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+
+/**
+ * Dialog for tool, that helps clean your storage from possibly now-unwanted videos
+ */
+public class CleanerDialog extends SkyTubeMaterialDialog {
+
+    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
+
+    public CleanerDialog(@NonNull MainActivity context) {
+        super(context);
+
+        title(R.string.cleaner);
+        // Set Layout for Dialog
+        customView(R.layout.dialog_clean_downloads,true);
+
+
+
+        positiveText(R.string.cleaner_confirm);
+        onPositive((dialog, action) -> {
+            boolean cleanWatchedDownloads = ((CheckBox)dialog.findViewById(R.id.clean_watched_downloads)).isChecked();
+            boolean cleanWatchedBookmarks = ((CheckBox)dialog.findViewById(R.id.clean_watched_bookmarks)).isChecked();
+            List<Single<Pair<Integer,Long>>> tasks = new ArrayList<>();
+            if (cleanWatchedDownloads) tasks.add(cleanWatchedDownloads(context));
+            if (cleanWatchedBookmarks) tasks.add(cleanWatchedBookmarks(context));
+            // MainFragment#refreshTabs
+            AtomicInteger removedVideosCnt = new AtomicInteger();
+            AtomicLong removedMB = new AtomicLong();
+            compositeDisposable.add(Single.merge(tasks).doFinally(() -> Toast.makeText(context, String.format(
+                    SkyTubeApp.getStr(R.string.cleaner_toast_finish), removedVideosCnt, removedMB), Toast.LENGTH_LONG).show())
+                    .subscribe(pair -> {
+                        removedVideosCnt.addAndGet(pair.first);
+                        removedMB.addAndGet(pair.second);
+            }));
+
+
+        });
+
+        cancelListener = dialog -> clearBackgroundTasks();
+    }
+
+    public void clearBackgroundTasks() { compositeDisposable.clear(); }
+
+    /**
+     * Cleans watched Videos, that are downloaded by the user
+     *
+     * @param context The Context for executing database actions and showing Toasts
+     * @return
+     */
+    private Single<Pair<Integer,Long>> cleanWatchedDownloads(@NonNull Context context) {
+        return Single.fromCallable(() -> {
+                    // Stats
+                    AtomicInteger removedVideosCnt = new AtomicInteger();
+                    AtomicLong removedB = new AtomicLong();
+                    // Maybe rewrite this with reactivex, so that it's parallel
+                    DownloadedVideosDb.getVideoDownloadsDb().getDownloadedVideos().stream()
+                            .filter((video) -> DownloadedVideosDb.getVideoDownloadsDb().isVideoDownloaded(video.getVideoId()).blockingGet()) // Filter only downloaded videos
```

Comment:
I'm not sure, I understand this. DownloadedVideosDb.getVideoDownloadsDb().getDownloadedVideos() returns a list of downloaded videos, why do you call the 'isVideoDownloaded' in the next line?

Reply:
Well, you are right, thats unnecessary